File: app/services/NifDispatcherActor.java
package services;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import com.fasterxml.jackson.databind.JsonNode;
import play.Logger;

import java.util.ArrayList;
import java.util.List;

/**
 * Akka actor handling a single network interface and forwards all arriving messages to all subscribing {@link WebSocketActor}. It also
 * cares for a WebSocket register where WebSocket actors can subscribe and unsubscribe.
 * <p>
 * It is created by the {@link PcapInitializer}.
 * <p>
 * Created by klange on 15.06.17.
 */
public class NifDispatcherActor extends UntypedActor {

    private static final Logger.ALogger LOGGER = Logger.of(NifDispatcherActor.class);

    public enum Protocol {REGISTER, UNREGISTER}

    private final String nifName;

    private List<ActorRef> webSocketRegister = new ArrayList();

    /**
     * Akka method to get this Actor started. Changes in props must be done in
     * the constructor too.
     */
    public static Props props(String nifName) {
        return Props.create(NifDispatcherActor.class, nifName);
    }

    public NifDispatcherActor(String nifName) {
        this.nifName = nifName;
    }

    @Override
    public void onReceive(final Object msg) throws Throwable {
        if(msg instanceof JsonNode) {
            webSocketRegister.forEach(actorRef -> actorRef.tell(msg, self()));
        } else if(Protocol.REGISTER.equals(msg)) {
            webSocketRegister.add(sender());
            LOGGER.info("registered WebSocket actor to network interface " + nifName);
        } else if(Protocol.UNREGISTER.equals(msg)) {
            webSocketRegister.remove(sender());
            LOGGER.info("unregistered WebSocket actor from network interface " + nifName);
        } else {
            unhandled(msg);
        }
    }
}

File: app/services/NifDispatcherActor.scala
package services

import akka.actor.{Actor, ActorRef, Props}
import play.api.Logger
import play.api.libs.json.JsObject
import services.NifDispatcherActor.{Subscribe, Unsubscribe}

import scala.collection.mutable.ArrayBuffer

/**
  * Akka actor handling a single network interface and forwards all arriving
  * messages to all subscribing [[WebSocketActor]]. It also cares for a
  * WebSocket register where WebSocket actors can subscribe and unsubscribe.
  *
  * A new NifDispatcherActor is created by the [[PcapInitializer]].
  *
  * Created by Kristian Lange on 2017.
  */
object NifDispatcherActor {

  def props(nifName: String) = Props(new NifDispatcherActor(nifName))

  case object Subscribe

  case object Unsubscribe

}

class NifDispatcherActor(nifName: String) extends Actor {

  private val logger: Logger = Logger(this.getClass)

  private val webSocketRegister: ArrayBuffer[ActorRef] = ArrayBuffer()

  def receive = {
    case msg: JsObject =>
      webSocketRegister.foreach(ws => ws ! msg)
    case Subscribe =>
      webSocketRegister += sender
      logger.info("subscribed WebSocket to network interface " + nifName)
    case Unsubscribe =>
      webSocketRegister -= sender
      logger.info("unsubscribed WebSocket from network interface " + nifName)
  }

}
